### PR TITLE
Jp itn

### PR DESCRIPTION
**Pull request by BuyuanCui**
# What does this PR do ?
Merge to main for Japanese support on cardinal, ordinal, time, date, fraction, decimal support
Add a one line overview of what this PR aims to accomplish.
This is a PR for Japanese ITN support.

# Before your PR is "Ready for review"
**Pre checks**:
- [*] Have you signed your commits? Use ``git commit -s`` to sign.
- [*] Do all unittests finish successfully before sending PR?
   1) ``pytest`` or (if your machine does not have GPU) ``pytest --cpu`` from the root folder (given you marked your test cases accordingly `@pytest.mark.run_only_on('CPU')`).
   2) Sparrowhawk tests ``bash tools/text_processing_deployment/export_grammars.sh --MODE=test ...``
- [*] If you are adding a new feature: Have you added test cases for both `pytest` and Sparrowhawk [here](tests/nemo_text_processing).
- [*] Have you added ``__init__.py`` for every folder and subfolder, including `data` folder which has .TSV files?
- [*] Have you followed codeQL results and removed unused variables and imports (report is at the bottom of the PR in github review box) ?
- [*] Have you added the correct license header `Copyright (c) 2023, NVIDIA CORPORATION & AFFILIATES. All rights reserved.` to all newly added Python files?
- [*] If you copied [nemo_text_processing/text_normalization/en/graph_utils.py](nemo_text_processing/text_normalization/en/graph_utils.py) your header's second line should be `Copyright 2015 and onwards Google, Inc.`. See an example [here](https://github.com/NVIDIA/NeMo-text-processing/blob/main/nemo_text_processing/text_normalization/en/graph_utils.py#L2).
- [*] Remove import guards (`try import: ... except: ...`) if not already done.
- [*] If you added a new language or a new feature please update the [NeMo documentation](https://github.com/NVIDIA/NeMo/blob/main/docs/source/nlp/text_normalization/wfst/wfst_text_normalization.rst) (lives in different repo).
- [*] Have you added your language support to [tools/text_processing_deployment/pynini_export.py](tools/text_processing_deployment/pynini_export.py).
  

**PR Type**:
- [*] New Feature
- [ ] Bugfix
- [ ] Documentation
- [ ] Test

If you haven't finished some of the above items you can still open "Draft" PR.

### CODE REVIEW

**Review by lazykyama — COMMENTED** [review]
I haven't reviewed each implementation detail due to a time constraint, but I checked all codes as much as possible. I'm not 100% sure the scope of this review, so please ignore my inappropriate comments.

Also let me put two more comments below.

In addition to my comments in `whitelist.tsv`, a role of this file is not clear for me. If we want to use whitelist to convert Katakana spoken representations for alphabetical abbreviated words (like PCIe, DGX, etc) into appropriate alphabet representations, we need to add more mappings.

And, regarding the interface of `inverse_normalize()`, as far as I checked, it requires a given input must be possible to be split by whitespaces into words/tokens. But, it's not good for Japanese language because a Japanese sentence usually doesn't include whitespaces as a delimiter between each word. If this function assumes that a given text includes one or more sentences, we need to improve the function interface or documentation, etc. (I guess the situation is the same as Mandarin.)

File: nemo_text_processing/inverse_text_normalization/jp/data/whitelist.tsv [hunk 1]
```diff
@@ -0,0 +1,257 @@
+PayPay	PayPay
```
[the remark this answers is not in the document]
Comment (BuyuanCui) [reply nemo_text_processing/inverse_text_normalization/jp/data/whitelist.tsv:1]: Thank you for pointing them out. I'll change them ASAP.

File: nemo_text_processing/inverse_text_normalization/jp/data/whitelist.tsv [hunk 2]
```diff
@@ -0,0 +1,257 @@
...
+Neet	ニート
+NG	エヌジー
+ng	エヌジー
+ol	オーエル
+OL	オーエル
+PV	ピーブイ
+pv	ピーブイ
+w杯	ワールドカップ
+w	わら
+W	わら
```
[the remark this answers is not in the document]
Comment (BuyuanCui) [reply nemo_text_processing/inverse_text_normalization/jp/data/whitelist.tsv]: I'd remove them.

File: nemo_text_processing/inverse_text_normalization/jp/data/whitelist.tsv [hunk 3]
```diff
@@ -0,0 +1,257 @@
+PayPay	PayPay
+ぺいぺい	PayPay
+AAA	トリプル a
+にってれ	日テレ
+ニッテレ	日テレ
+COVID-19	コビット 19
+cm	シーエム
+CM	シーエム
+GW	ジーダブル
+gw	シーダブル
```
[the remark this answers is not in the document]
Comment (BuyuanCui) [reply nemo_text_processing/inverse_text_normalization/jp/data/whitelist.tsv]: Would apply updates accordingly.

File: tests/nemo_text_processing/jp/data_inverse_text_normalization/test_cases_cardinal.txt [hunk 4]
```diff
@@ -0,0 +1,103 @@
+一~1
+百~100
+五万億~5,000,000,000,000
+五千億~500,000,000,000
+五万~50,000
+五兆~5,000,000,000,000
+十一兆~11,000,000,000,000
+十一兆一~11,000,000,000,001
+九十九兆~99,000,000,000,000
+いっちょう~1,000,000,000,000
```
[the remark this answers is not in the document]
Comment (BuyuanCui) [reply tests/nemo_text_processing/jp/data_inverse_text_normalization/test_cases_cardinal.txt:9]: This is really helpful! I should remove it.

File: tests/nemo_text_processing/jp/data_inverse_text_normalization/test_cases_date.txt [hunk 5]
```diff
@@ -0,0 +1,21 @@
+一日~1日
```
[the remark this answers is not in the document]
Comment (BuyuanCui) [reply tests/nemo_text_processing/jp/data_inverse_text_normalization/test_cases_date.txt:1]: I would bring to discussion with the team. This is very helpful!

[the remark this answers is not in the document]
Comment (BuyuanCui) [reply tests/nemo_text_processing/jp/data_inverse_text_normalization/test_cases_date.txt:1]: After a discussion, I decided that I would only support one form, '1日'. The reason being that the type of many-to-one mapping is not supported. I created a Jira task to address this issue in the future. For the initial release we have, I would have to push this issue back. Below is the link for Jira I created.   
[6989](https://jirasw.nvidia.com/browse/JPE-6989)

File: tests/nemo_text_processing/jp/data_inverse_text_normalization/test_cases_cardinal.txt [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+五万~50,000
+五兆~5,000,000,000,000
+十一兆~11,000,000,000,000
+十一兆一~11,000,000,000,001
+九十九兆~99,000,000,000,000
+いっちょう~1,000,000,000,000
+いっちょうせんおく~1,100,000,000,000
+いっちょうにせんおく~1,200,000,000,000
+いっちょうごせんひゃくじゅうおくいち~1,511,000,000,001
+いっちょうごせんじゅうおくせん~1,501,000,001,000
```
[the remark this answers is not in the document]
Comment (BuyuanCui) [reply tests/nemo_text_processing/jp/data_inverse_text_normalization/test_cases_cardinal.txt]: I have been thinking about this and thought I might need to have another grammar for this issue. This cardinal grammar is only responsible for pure numeric outputs. Will bring discuss with team.

File: nemo_text_processing/inverse_text_normalization/jp/taggers/whitelist.py [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+class WhiteListFst(GraphFst):
+    """
+    Finite state transducer for classifying whitelisted tokens
+    This class has highest priority among all classifier grammars. Whitelisted tokens are defined and loaded from "data/whitelist.tsv".
+    """
+
+    def __init__(self, input_case: str = INPUT_LOWER_CASED, input_file: str = None):
+        super().__init__(name="whitelist", kind="classify")
+
+        whitelist = pynini.string_file(get_abs_path("data/whitelist.tsv"))
```
[the remark this answers is not in the document]
Comment (lazykyama) [reply nemo_text_processing/inverse_text_normalization/jp/taggers/whitelist.py:32]: @BuyuanCui Let me confirm if you don't want to add warning message or not. If no warn message is necessary here, please ignore this comment.

[the remark this answers is not in the document]
Comment (BuyuanCui) [reply nemo_text_processing/inverse_text_normalization/jp/taggers/whitelist.py:32]: Thank you for the comment! I'll remove the <input_file>.

**Comment by lazykyama** [comment]
@BuyuanCui I reviewed the latest codes and it's almost OK. But, I'm so sorry that it looks like I missed one more thing.

So far, the package and directory names include `jp` like `nemo_text_processing/inverse_text_normalization/jp/__init__.py`, but I think it would be better to be the same with a standard language code (https://en.wikipedia.org/wiki/List_of_ISO_639-1_codes). Would it be necessary to follow this standard? If yes, it should be changed to `ja`. If not, please ignore this.

So sorry again for this late comment. There's no additional thing except for this. Thanks.

**Review by lazykyama — COMMENTED** [review]
@BuyuanCui Sorry for my late comment. Almost looks good to me! But, the last thing is `nemo_text_processing/inverse_text_normalization/ja/taggers/Untitled.ipynb`. I think it was accidentally submitted. Is that right? If yes, please remove it from this PR.

**Comment by ekmb** [comment]
@BuyuanCui should we close this PR in favor of https://github.com/NVIDIA/NeMo-text-processing/pull/141 ?

**Pull request closed without merging**